libs/api_common/windows_platform_common.cpp

```diff
@@ -6,60 +6,351 @@
+        EBPF_RETURN_VOID();
+    }
+
+    ebpf_free((void*)descriptor->context_descriptor);
+    ebpf_free((void*)descriptor->platform_specific_data);
+
+    delete descriptor;
+
+    EBPF_RETURN_VOID();
+}
+
+struct EbpfProgramType_deleter
+{
+    void
+    operator()(_In_ _Post_invalid_ EbpfProgramType* descriptor)
+    {
+        _ebpf_program_descriptor_free(descriptor);
+    }
+};
+
+// Thread local cache for program information queried from execution context.
+typedef std::unique_ptr<ebpf_program_info_t, _ebpf_program_info_deleter> ebpf_program_info_ptr_t;
+static thread_local std::map<ebpf_program_type_t, ebpf_program_info_ptr_t, guid_compare> _program_info_cache;
+
+// Thread local cache for program descriptor queried from execution context.
+typedef std::unique_ptr<EbpfProgramType, EbpfProgramType_deleter> ebpf_program_descriptor_ptr_t;
+static thread_local std::map<ebpf_program_type_t, ebpf_program_descriptor_ptr_t, guid_compare>
+    _program_descriptor_cache;
+
+// Global cache for the program and section information queried from eBPF store.
+static std::map<ebpf_program_type_t, EbpfProgramType, guid_compare> _windows_program_types;
+static std::vector<ebpf_section_definition_t> _windows_section_definitions;
+static std::map<ebpf_program_type_t, ebpf_program_info_t, guid_compare> _windows_program_information;
+
+void
+set_program_under_verification(ebpf_handle_t program)
+{
+    _program_under_verification = program;
+}
+
+ebpf_result_t
+_get_program_descriptor_from_info(_In_ const ebpf_program_info_t* info, _Outptr_ EbpfProgramType** descriptor)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    EbpfProgramType* type = nullptr;
+
+    try {
+        char* name = nullptr;
+        type = new (std::nothrow) EbpfProgramType();
+        if (type == nullptr) {
+            result = EBPF_NO_MEMORY;
+            goto Exit;
+        }
+
+        name = _strdup(info->program_type_descriptor.name);
+        if (name == nullptr) {
+            result = EBPF_NO_MEMORY;
+            goto Exit;
+        }
+        type->name = std::string(name);
+        type->context_descriptor = (ebpf_context_descriptor_t*)ebpf_allocate(sizeof(ebpf_context_descriptor_t));
+        if (type->context_descriptor == nullptr) {
+            result = EBPF_NO_MEMORY;
+            goto Exit;
+        }
+        memcpy(
+            (void*)type->context_descriptor,
+            info->program_type_descriptor.context_descriptor,
+            sizeof(ebpf_context_descriptor_t));
+        ebpf_program_type_t* program_type = (ebpf_program_type_t*)ebpf_allocate(sizeof(ebpf_program_type_t));
+        if (program_type == nullptr) {
+            result = EBPF_NO_MEMORY;
+            goto Exit;
+        }
+        *program_type = info->program_type_descriptor.program_type;
+        type->platform_specific_data = (uint64_t)program_type;
+        type->is_privileged = info->program_type_descriptor.is_privileged;
+
+        *descriptor = type;
+    } catch (...) {
```

Comment:
Is there a coding guidance around this? I see several instances of such catch-all in ebpf-api and elsewhere? Why cant specific exceptions be handled instead?

Reply:
These functions cannot throw, so we have to catch all the exceptions. Regarding masking AVs, I think we cannot catch AVs unless we compile with SEH option, which we are not.

https://docs.microsoft.com/en-us/cpp/cpp/structured-exception-handling-c-cpp?view=msvc-160